jsonnet/exclude-rules.libsonnet

```diff
@@ -1,54 +1,93 @@
+local excludeRule(rule, excludedRules) =
+  if std.length(excludedRules) == 0 then
+    [rule]
+  else if (('alert' in rule && 'alert' in excludedRules[0]) && rule.alert == excludedRules[0].alert) ||
+          (('record' in rule && 'record' in excludedRules[0]) && rule.record == excludedRules[0].record) then
```

Comment:
@dgrisonnet in https://github.com/openshift/cluster-monitoring-operator/pull/1076/commits/362dc973bdf0fa03478df49c2364e5d3e8ba8d48#diff-9a529e7399b36b3c02f816e864690cfad2559b40127f86268cc44c5dbce1277fR47 i need a `startsWith` comparator.

How feasible is it to to extend this function that enables that as well?

cc @paulfantom

Reply:
would it be meaningful to use regex here? e.g. https://tanka.dev/jsonnet/native#regexmatch

I am not familiar at all with `std.native`, what do you think about it @paulfantom?

Reply:
`std.native` is a jsonnet extension from tanka, so we would need to change more things just to have this function.

Reply:
I think we could merge `excludeRule` and `patchRule` as those are functionally almost the same and it would be possible to add startsWith there too.

Something like:
```
local patchOrExcludeRule(rule, ruleSet, operation) = 
  if std.length(ruleSet) == 0 then
    [rule]
  else if (('alert' in rule && 'alert' in ruleSet[0]) && rule.alert == ruleSet[0].alert) ||
          (('record' in rule && 'record' in ruleSet[0]) && rule.record == ruleSet[0].record) then
    if operation == 'patch' then
      [std.mergePatch(rule, patchedRules[0])]
    else
      []
  else
    [] + patchOrRemoveRule(rule, ruleSet[1:]);
```
Same would apply to `excludeRuleGroup` and `patchRuleGroup`.

As for the `startsWith` operator, maybe we can do a naive match and replace `rule.alert == ruleSet[0].alert` with `startsWith` ?

Reply:
:100: I'll refactor all the recursive functions based on your suggestion.

> As for the startsWith operator, maybe we can do a naive match and replace rule.alert == ruleSet[0].alert with startsWith ?

:+1: I don't think we need more than that